keps/sig-api-machinery/4933-cacher-label-index/README.md

```diff
@@ -0,0 +1,916 @@
+bogged down.
+-->
+
+#### Story 1
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+Since this feature is relatively large, the entire feature will be controlled by a feature gate `CacherLabelIndex` in the early stage.
+
+Considering that daemonset and spark applications coexist in a cluster, and they watch based on `spec.nodeName` and `applicationId` label respectively, we will support both usage at the same time.
+
+In reality, a single filter condition, such as nodeName or applicationId, can already accurately match cacheWatcher, this feature still considers supporting only one triggerIndex for a watch request. At the same time, since the following prerequisites guarantee that OR operations are not supported, filtering events with a single triggerIndex can ensure correctness
+- Label selectors and field selectors in a watch request are AND operations.
+- Multiple requirements in label selector are also AND operations.
+- Through a [`RequiresExactMatch`](https://github.com/kubernetes/kubernetes/blob/v1.31.2/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L574) check, a set-based label selector or field selector that supports OR operations on multiple values is considered not matching trigger indexes in cacher.
+
+### Commandline changes
+
+Considering that label names may be different in different scenarios, such as app label name in Flink and spark-app-selector label name in Spark, this feature will add a commandline parameter to the apiserver, allowing different label names to be configured as the trigger index of the apiserver cacher.
+
+```
+--index-labels. Index labels settings for some resources (pods, nodes, etc.), comma separated. The individual setting format: resource[.group]#labelName1;labelName2, where resource is lowercase plural (no version), group is omitted for resources of apiVersion v1 (the legacy core API) and included for others, and labelName is a string. It takes effect when watch-cache and feature gate CacherLabelIndex are enabled. In addition to built-in indexes such as pod.spec.nodeName, apiserver cacher will create additional indexes for these labels to speed up list and watch requests.
+
+# e.g. for spark
+--index-labels=pods#spark-app-selector
+# e.g. for spark and flink
+--index-labels=pods#spark-app-selector;app
```

Comment:
Do we need the `;` support?  I'm tempted to keep it simple and just require `pods#spark-app-selector,pods#app`.

While I do agree that it's more compact to offer the `;`, it also means there are multiple ways to express the same thing, and it doesn't seem all the necessary, so I'd rather use the simplest grammar, even if it's slightly more verbose.

Reply:
I updated this to use comma separator and to add resources in each separated item ```--index-labels=pods#spark-app-selector,pods#app,nodes#zone```